File: admin-panel/src/services/courseService.ts
import api from './api';
// import { mockDataService } from './mockDataService';
import type { PaginatedResponse, SearchParams, PaginationParams } from '../types/api';
import type { Course, Enrollment } from '../types/course';
import { CourseStatus, CourseLevel } from '../types/course';

export interface CourseStatistics {
  totalCourses: number;
  activeCourses: number;
  publishedCourses: number;
  draftCourses: number;
  totalEnrollments: number;
  averageCompletionRate: number;
  averageRating: number;
  recentCourses: number;
  courseGrowthRate: number;
}

export interface CreateCourseRequest {
  title: string;
  description: string;
  shortDescription?: string;
  level: CourseLevel;
  price: number;
  duration: number; // in hours
  maxStudents?: number;
  prerequisites?: string[];
  learningObjectives?: string[];
  syllabus?: string;
  thumbnail?: string;
  teacherId: string;
}

export interface UpdateCourseRequest {
  title?: string;
  description?: string;
  shortDescription?: string;
  level?: CourseLevel;
  price?: number;
  duration?: number;
  maxStudents?: number;
  prerequisites?: string[];
  learningObjectives?: string[];
  syllabus?: string;
  thumbnail?: string;
  status?: CourseStatus;
}

export interface CourseProgress {
  courseId: string;
  courseName: string;
  totalStudents: number;
  completedStudents: number;
  averageProgress: number;
  completionRate: number;
  averageRating: number;
  totalRatings: number;
}

class CourseService {
  // Get course statistics
  async getCourseStatistics(): Promise<CourseStatistics> {
    try {
      const response = await api.get('/courses/statistics');
      return response.data;
    } catch (error) {
      console.error('Failed to fetch course statistics:', error);
      // Return empty statistics instead of mock data
      return {
        totalCourses: 0,
        activeCourses: 0,
        publishedCourses: 0,
        draftCourses: 0,
        totalEnrollments: 0,
        averageCompletionRate: 0,
        averageRating: 0,
        recentCourses: 0,
        courseGrowthRate: 0,
      };
    }
  }

  // Get all courses with pagination and filtering
  async getCourses(params: SearchParams & {
    status?: CourseStatus;
    level?: CourseLevel;
    teacherId?: string;
    minPrice?: number;
    maxPrice?: number;
  } = {}): Promise<PaginatedResponse<Course>> {
    try {
      const response = await api.get('/courses', { params });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch courses:', error);
      // Return empty paginated response instead of mock data
      return {
        content: [],
        totalElements: 0,
        totalPages: 0,
        size: 10,
        number: 0,
        first: true,
        last: true,
        numberOfElements: 0,
        empty: true,
      };
    }
  }

  // Get course by ID
  async getCourseById(id: string): Promise<Course> {
    const response = await api.get(`/courses/${id}`);
    return response.data;
  }

  // Create new course
  async createCourse(courseData: CreateCourseRequest): Promise<Course> {
    const response = await api.post('/courses', courseData);
    return response.data;
  }

  // Update course
  async updateCourse(id: string, courseData: UpdateCourseRequest): Promise<Course> {
    const response = await api.put(`/courses/${id}`, courseData);
    return response.data;
  }

  // Delete course
  async deleteCourse(id: string): Promise<void> {
    await api.delete(`/courses/${id}`);
  }

  // Search courses
  async searchCourses(params: SearchParams & {
    status?: CourseStatus;
    level?: CourseLevel;
    teacherId?: string;
    minPrice?: number;
    maxPrice?: number;
  } = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get('/courses/search', { params });
    return response.data;
  }

  // Get courses by status
  async getCoursesByStatus(status: CourseStatus, params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get(`/courses/status/${status}`, { params });
    return response.data;
  }

  // Get courses by level
  async getCoursesByLevel(level: CourseLevel, params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get(`/courses/level/${level}`, { params });
    return response.data;
  }

  // Get courses by teacher
  async getCoursesByTeacher(teacherId: string, params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get(`/courses/teacher/${teacherId}`, { params });
    return response.data;
  }

  // Get courses by student
  async getCoursesByStudent(studentId: string, params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get(`/courses/student/${studentId}`, { params });
    return response.data;
  }

  // Get published courses
  async getPublishedCourses(params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get('/courses/published', { params });
    return response.data;
  }

  // Get recent courses
  async getRecentCourses(params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get('/courses/recent', { params });
    return response.data;
  }

  // Get popular courses
  async getPopularCourses(params: PaginationParams = {}): Promise<PaginatedResponse<Course>> {
    const response = await api.get('/courses/popular', { params });
    return response.data;
  }

  // Update course status
  async updateCourseStatus(id: string, status: CourseStatus): Promise<Course> {
    const response = await api.patch(`/courses/${id}/status`, null, {
      params: { status }
    });
    return response.data;
  }

  // Publish course
  async publishCourse(id: string): Promise<Course> {
    const response = await api.patch(`/courses/${id}/publish`);
    return response.data;
  }

  // Unpublish course
  async unpublishCourse(id: string): Promise<Course> {
    const response = await api.patch(`/courses/${id}/unpublish`);
    return response.data;
  }

  // Archive course
  async archiveCourse(id: string): Promise<Course> {
    const response = await api.patch(`/courses/${id}/archive`);
    return response.data;
  }

  // Get course progress
  async getCourseProgress(params: PaginationParams = {}): Promise<CourseProgress[]> {
    try {
      const response = await api.get('/courses/progress', { params });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch course progress:', error);
      // Return empty array instead of mock data
      return [];
    }
  }

  // Get course enrollments
  async getCourseEnrollments(courseId: string, params: PaginationParams = {}): Promise<PaginatedResponse<Enrollment>> {
    const response = await api.get(`/courses/${courseId}/enrollments`, { params });
    return response.data;
  }

  // Check user access to course
  async checkUserAccess(courseId: string, userId: string): Promise<boolean> {
    const response = await api.get(`/courses/${courseId}/access/${userId}`);
    return response.data;
  }
}

export const courseService = new CourseService();
export default courseService;

File: backend/src/main/kotlin/com/essenceqa/lms/service/CourseService.kt
package com.essenceqa.lms.service

import com.essenceqa.lms.model.Course
import com.essenceqa.lms.model.Teacher
import com.essenceqa.lms.repository.CourseRepository
import com.essenceqa.lms.repository.TeacherRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDateTime

@Service
@Transactional
class CourseService(
    private val courseRepository: CourseRepository,
    private val teacherRepository: TeacherRepository
) {
    
    // Get all courses with pagination
    fun getAllCourses(pageable: Pageable): Page<Course> {
        return courseRepository.findActiveCourses(pageable)
    }
    
    // Get course by ID
    fun getCourseById(id: String): Course? {
        return courseRepository.findById(id).orElse(null)
    }
    
    // Create new course
    fun createCourse(course: Course): Course {
        return courseRepository.save(course)
    }
    
    // Create new course with parameters
    fun createCourse(
        title: String,
        description: String?,
        teacherId: String,
        startDate: LocalDateTime,
        endDate: LocalDateTime,
        meetingLink: String?,
        price: BigDecimal,
        maxStudents: Int?,
        isBrowseable: Boolean
    ): Course {
        val teacher = teacherRepository.findById(teacherId)
            .orElseThrow { IllegalArgumentException("Teacher not found with id: $teacherId") }
        
        val course = Course(
            title = title,
            description = description,
            teacherId = teacherId,
            teacher = teacher,
            startDate = startDate,
            endDate = endDate,
            meetingLink = meetingLink,
            price = price,
            maxStudents = maxStudents,
            isActive = true,
            isBrowseable = isBrowseable
        )
        
        return courseRepository.save(course)
    }
    
    // Update course
    fun updateCourse(id: String, updatedCourse: Course): Course? {
        return if (courseRepository.existsById(id)) {
            courseRepository.save(updatedCourse.copy(id = id))
        } else {
            null
        }
    }
    
    // Delete course (soft delete by setting isActive to false)
    fun deleteCourse(id: String): Boolean {
        return courseRepository.findById(id).map { course ->
            courseRepository.save(course.copy(isActive = false))
            true
        }.orElse(false)
    }
    
    // Search courses
    fun searchCourses(search: String, pageable: Pageable): Page<Course> {
        return if (search.isBlank()) {
            courseRepository.findActiveCourses(pageable)
        } else {
            courseRepository.searchCourses(search, pageable)
        }
    }
    
    // Get courses by teacher
    fun getCoursesByTeacher(teacherId: String, pageable: Pageable): Page<Course> {
        return courseRepository.findByTeacherIdAndIsActive(teacherId, true, pageable)
    }
    
    // Get courses by student (enrolled courses)
    fun getCoursesByStudent(studentId: String, pageable: Pageable): Page<Course> {
        return courseRepository.findCoursesByStudentId(studentId, pageable)
    }
    
    // Get published courses
    fun getPublishedCourses(pageable: Pageable): Page<Course> {
        return courseRepository.findBrowseableCourses(pageable)
    }
    
    // Get recent courses (last 30 days)
    fun getRecentCourses(pageable: Pageable): Page<Course> {
        val thirtyDaysAgo = LocalDateTime.now().minusDays(30)
        return courseRepository.findRecentCourses(thirtyDaysAgo, pageable)
    }
    
    // Get courses by price range
    fun getCoursesByPriceRange(minPrice: BigDecimal, maxPrice: BigDecimal, pageable: Pageable): Page<Course> {
        return courseRepository.findByPriceRange(minPrice, maxPrice, pageable)
    }
    
    // Publish course
    fun publishCourse(id: String): Course? {
        return courseRepository.findById(id).map { course ->
            courseRepository.save(course.copy(isBrowseable = true))
        }.orElse(null)
    }
    
    // Unpublish course
    fun unpublishCourse(id: String): Course? {
        return courseRepository.findById(id).map { course ->
            courseRepository.save(course.copy(isBrowseable = false))
        }.orElse(null)
    }
    
    // Archive course
    fun archiveCourse(id: String): Course? {
        return courseRepository.findById(id).map { course ->
            courseRepository.save(course.copy(isActive = false))
        }.orElse(null)
    }
    
    // Get course statistics
    fun getCourseStatistics(): Map<String, Any> {
        val totalCourses = courseRepository.countActiveCourses()
        val publishedCourses = courseRepository.countPublishedCourses()
        val draftCourses = courseRepository.countDraftCourses()
        val totalEnrollments = courseRepository.countTotalEnrollments()
        val thirtyDaysAgo = LocalDateTime.now().minusDays(30)
        val recentCourses = courseRepository.countRecentCourses(thirtyDaysAgo)
        
        return mapOf(
            "totalCourses" to totalCourses,
            "activeCourses" to totalCourses,
            "publishedCourses" to publishedCourses,
            "draftCourses" to draftCourses,
            "totalEnrollments" to totalEnrollments,
            "averageCompletionRate" to 0.0, // TODO: Calculate from actual data
            "averageRating" to 0.0, // TODO: Calculate from actual data
            "recentCourses" to recentCourses,
            "courseGrowthRate" to 0.0 // TODO: Calculate growth rate
        )
    }
    
    // Get course progress (placeholder for now)
    fun getCourseProgress(): List<Map<String, Any>> {
        // TODO: Implement actual course progress calculation
        return emptyList()
    }
}